### PR TITLE
Add `parseSignature` functions to ABI utils

### PR DESCRIPTION
**Pull request by haltman-at**
## PR description

This PR adds functions to ABI utils for parsing signatures.  There's `parseFunctionSignature`, `parseEventSignature`, and `parseErrorSignature`, depending on the entry type you want out.  There's also just a bare-bones `parseSignature` that doesn't return a valid ABI entry, but which is used to back the other three.

This is intended as precursor work for our upcoming solution to #6029.

Note that the ABI entries generated by these functions will have default values for lots of fields... so e.g. `anonymous` will always be `false`, `outputs` will always be `[]`, etc.  Best we can do in the situation, seems to me.  It won't be relevant in the contexts we need this (#6029) anyway.

Also note that the parser doesn't fully validate the input... when it came to primitive types, I just checked that the types looked vaguely like primitive types, not that they're actual types.  (So e.g. `kafoo2e3` will pass validation.)  I could tighten this up if you think it's warranted.

(Also note that on the other hand, if you're dealing with a weird library signature that has like periods or spaces in it, that will *not* pass validation.  I'd consider that to be out of scope for this though... `abi-utils` deals in valid ABIs, not the weird stuff libraries sometimes do.  And of course for #6029 we should only need valid signatures.)

Anyway, I'm not going to bother trying to explain here how the parser works... just go read the code if you want to know that. :P

Also I just used ordinary `Error`s for when something went wrong, I didn't see any point in using like a custom `Error` class here, but again let me know if you think that should be changed.

## Testing instructions

I think the tests I added -- particularly the property-based tests -- should do the job pretty well! :)

Note: Turns out that TS doesn't recognize the arbitraries for `FunctionEntry`, `EventEntry`, etc, as being of the appropriate types. :-/  Maybe there's a way to fix that, but I don't really know much about arbitraries, so I just tossed in some coercions instead...

### CODE REVIEW

File: packages/abi-utils/lib/parse.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  const parameterList = signature.slice(openParenIndex);
+  const name = signature.slice(0, openParenIndex);
+  const inputs = parseParameterList(parameterList);
+  return {
+    name,
+    inputs
+  };
+}
+
+export function parseParameterList(parameterList: string): Parameter[] {
```
Comment (gnidan) [inline packages/abi-utils/lib/parse.ts:50]: Does this (and the other parameter list one) need to be exported? Or are these more internal concerns?

File: packages/abi-utils/lib/index.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 export * from "./types";
 export * from "./normalize";
 export * from "./signature";
+export * from "./fromSignature";
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/abi-utils/lib/index.ts]: Ooh good point.  (Also ha you started this review before I fixed that mistake huh? :P )

File: packages/abi-utils/lib/parse.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  const { parameters, remaining } =
+    parseParameterListWithRemainder(parameterList);
+  if (remaining !== "") {
```
Comment (cds-amal) [inline packages/abi-utils/lib/parse.ts:53]: Is there a case to allow trailing spaces?

Comment (haltman-at) [reply packages/abi-utils/lib/parse.ts:53]: I don't see any reason to.

File: packages/abi-utils/lib/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/abi-utils/lib/index.ts]: OK, did this!

File: packages/abi-utils/lib/parse.ts (hunk 1 above)
Comment (haltman-at) [reply packages/abi-utils/lib/parse.ts:50]: Removed export on parse-with-remainder; left it on on the others.

**Comment by gnidan** [comment]
Note we could probably use the ABI schema inside contract-schema to do the validation pretty easily! (Should probably open an issue for this also)

**Pull request merged**